### PR TITLE
Optimize .travis.yml to reduce network calls and use build stages

### PR DESCRIPTION
**Pull request by dimroc**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,38 +1,51 @@
...
-  - sudo apt-get install -y -qq yarn
+  # Code climate
   - curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > $GOPATH/bin/cc-test-reporter
   - chmod +x $GOPATH/bin/cc-test-reporter
 
-install: ./internal/ci/install_dependencies
-
-script:
-  - bash $COMMAND
+jobs:
```
Comment (rupurt) [inline .travis.yml]: 👍 this is a lot easier to understand

File: .travis.yml [hunk 2]
```diff
@@ -1,38 +1,51 @@
...
+  include:
+    - stage: Test
+      name: "Go Tests"
+      script:
+        - dep ensure
```
Comment (j16r) [inline .travis.yml]: Does this mean that go tests and gui tests both have to run `dep ensure`? What if they get different versions of dependencies? go dep's solver is not deterministic IIRC.

File: .travis.yml [hunk 3]
```diff
@@ -1,38 +1,51 @@
...
+        - ./internal/ci/go_test
```
Comment (rupurt) [inline .travis.yml]: If this doesn't do a `yarn install` won't it fail during the build process?

**Comment by rupurt** [comment]
@dimroc I wonder if it's also worthwhile setting the registry to npmjs e.g.

```
yarn config set registry http://registry.npmjs.org/
```

File: .travis.yml (hunk 2 above)
Comment (dimroc) [reply .travis.yml]: No, they're broken up per script and run in parallel. Each entry in the yaml array (marked by `- <something>`) is run in parallel unless it's a new stage, but we only have one stage: test. So GUI tests only run `yarn install`, while this only runs `dep ensure`.

Furthermore, `go dep` uses `Gopkg.lock` to lock dependencies to specific SHAs so it should get the same versions regardless.

**Comment by dimroc** [comment]
I haven't seen that done in any .travis.yml file examples so I'm leaning towards no. Let me know if you feel strongly about it.

**Comment by rupurt** [comment]
@dimroc I don't feel super strongly. The yarnpkg registry should just be a mirror of the npm registry. If we do get request errors it's probably worth checking both status pages to see if there's a known issue.

https://status.yarnpkg.com/
https://status.npmjs.org/

File: .travis.yml (hunk 3 above)
Comment (rupurt) [reply .travis.yml]: Oh, my bad. I was thinking it builds the assets. But that doesn't happen with the tests.

**Comment by rupurt** [comment]
🏀

**Pull request merged**